### ISSUE TITLE
Arris CM820A modem - invalid literal error

### ISSUE THREAD

**Issue by shadowpuck99** [twstokes/arris-scrape#1]
Hi - this is really cool and for some reason, I just can't get it working....

anytime i try to run this, the following errors pop up:

2021-10-04T20:37:59.033366720Z Modem scraper running.,
2021-10-04T20:37:59.033399821Z Error: (<class 'ValueError'>, ValueError("invalid literal for int() with base 10: 'None'"), <traceback object at 0x7fc2e84b4080>),
2021-10-04T20:37:59.033404821Z Error: (<class 'ValueError'>, ValueError("invalid literal for int() with base 10: 'None'"), <traceback object at 0x7fc2e8485c80>),
2021-10-04T20:37:59.033408021Z Error: (<class 'ValueError'>, ValueError("invalid literal for int() with base 10: 'None'"), <traceback object at 0x7fc2e84b2cc0>),
2021-10-04T20:37:59.033411021Z Error: (<class 'ValueError'>, ValueError("invalid literal for int() with base 10: 'None'"), <traceback object at 0x7fc2e848acc0>),
2021-10-04T20:37:59.033414021Z Error: (<class 'ValueError'>, ValueError("invalid literal for int() with base 10: 'None'"), <traceback object at 0x7fc2e84b2cc0>),
2021-10-04T20:37:59.033417021Z Abort! Max retries reached: 5

my modem status page looks identical to your screenshot.  i'm certainly missing something simple here, but, can't seem to figure it out.  would you be willing to take a quick look?

Thank you,

**Comment by twstokes** [comment]
Hello 👋! I don't think I did a very good job of outputting exception info.

I've added this [traceback branch](https://github.com/twstokes/arris-scrape/tree/traceback) which may improve things. Could you switch to that branch and rerun? If using Docker, the container will need to be rebuilt.

**Comment by shadowpuck99** [comment]
hmmm - i should also point out i'm fairly new to docker.....so i apologize if i'm doing something totally silly here!

i started over completely with a new vm just to try and eliminate any other possibilities....

now, when i try to run it here's what the logs show:
nfluxdb (this just seems to run nonstop in the log, or a line that is very similar)
[httpd] 172.18.0.3 - root [05/Oct/2021:03:15:33 +0000] "POST /write?db=modem HTTP/1.1 " 204 0 "-" "python-requests/2.26.0" 80e26f30-258a-11ec-86c5-0242ac120002 125

scrape
No log line matching the '' filter

**Comment by shadowpuck99** [comment]
actually, thought it might be better if i just posted some logs from each item (scraper, grafana, and influxdb) when using the traceback branch....
[_arrisscrape_grafana_1_logs.txt](https://github.com/twstokes/arris-scrape/files/7286635/_arrisscrape_grafana_1_logs.txt)
[_arrisscrape_influxdb_1_logs.txt](https://github.com/twstokes/arris-scrape/files/7286636/_arrisscrape_influxdb_1_logs.txt)
[_arrisscrape_modem_scraper_1_logs.txt](https://github.com/twstokes/arris-scrape/files/7286637/_arrisscrape_modem_scraper_1_logs.txt)

**Comment by twstokes** [comment]
This error is telling us that when it tried to connect to the InfluxDB database, it failed to connect to a host named "influxdb".

```
urllib3.exceptions.MaxRetryError: HTTPConnectionPool(host='influxdb', port=8086): Max retries exceeded with url: /write?db=modem (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7f4975c93040>: Failed to establish a new connection: [Errno 111] Connection refused'))
```

The "Full Service" Docker Compose config defines that hostname [here](https://github.com/twstokes/arris-scrape/blob/master/full-service.yml#L9).

So to make sure I understand your setup:

1. Are you running Option A or Option B [here](https://github.com/twstokes/arris-scrape/blob/master/README.md)?
2. If Option B, what does your `config.py` look like?

**Comment by shadowpuck99** [comment]
i am running Option B.
i was having different issues the first time i tried with the master branch.
so, i wonder what i missed here - as i said, i built a new vm and started over for this.
if i'm understanding things right i should be able to see http://localhost:8086 if the influxdb container is running, right?  if so, i can't and that's definitely a problem!

"""
Parameters for the modem scraper.
"""
scraper_config = {
    'modem_url': 'http://192.168.100.1', # POPULATE MODEM URL HERE
    'max_retries': 5,
    'poll_interval_seconds': 30
}

"""
Parameters for InfluxDBClient. See influxdb.InfluxDBClient for all options.
"""
influx_config = {
    'host': 'influxdb', # POPULATE INFLUX_DB HOST HERE
    'port': 8086,
    'database': 'modem'
}

**Comment by twstokes** [comment]
> if i'm understanding things right i should be able to see http://localhost:8086 if the influxdb container is running, right?

You probably won't be able to see this even when everything's working since InfluxDB isn't exposed outside of the container.

What's your output of `docker-compose ps` when inside this branch? You may want to do a `docker-compose down -v --remove-orphans` to wipe everything clean and restart.

Your config looks good, though my modem URL is `http://192.168.100.1/cgi-bin/status_cgi`. However, the errors in your recent log are more about the script not being able to talk to the `influxdb` host.

**Comment by shadowpuck99** [comment]
thank you, i did realize it wasn't exposed after i asked the question!  :)

here's the output:
           Name                   Command        State    Ports
---------------------------------------------------------------
arrisscrape_modem_scraper_1   python scrape.py   Exit 1        

at the moment, it's not running - i can restart it and rerun that command if needed...

previously - before starting over i had tried the URL you listed and i wasn't getting any data then either...

docker-compose down -v --remove-orphans
Removing orphan container "arrisscrape_influxdb_1"
Removing arrisscrape_modem_scraper_1 ... done
Removing network arrisscrape_default

ok - just to confirm - i would now start over at step 3 of option b?

thank you again for taking the time to help me out...

**Comment by twstokes** [comment]
> ok - just to confirm - i would now start over at step 3 of option b?

Correct!

> thank you again for taking the time to help me out...

No problem!

**Comment by shadowpuck99** [comment]
hmmm....given my lack of experience with docker i've probably missed something somewhere.

it did the same thing after starting over.

i wiped out the containers and images again and went back to the master branch.

the behavior is now same as i had prior - meaning, it tries 5 times and then quits....
Modem scraper running.,
Error: (<class 'ValueError'>, ValueError("invalid literal for int() with base 10: 'None'"), <traceback object at 0x7fa2f6044ac0>),
Error: (<class 'ValueError'>, ValueError("invalid literal for int() with base 10: 'None'"), <traceback object at 0x7fa2f601b080>),
Error: (<class 'ValueError'>, ValueError("invalid literal for int() with base 10: 'None'"), <traceback object at 0x7fa2f6042380>),
Error: (<class 'ValueError'>, ValueError("invalid literal for int() with base 10: 'None'"), <traceback object at 0x7fa2f601bd40>),
Error: (<class 'ValueError'>, ValueError("invalid literal for int() with base 10: 'None'"), <traceback object at 0x7fa2f6039700>),
Abort! Max retries reached: 5

so, seems like there's something i'm doing it doesn't like!  i'll have to some more digging/reading on my end to see what i'm missing or not doing right here....ahh, computers are fun....

**Comment by twstokes** [comment]
I was hoping with the new branch the traceback would be more verbose and we could see exactly where it's dying. My best guess is that it's trying to parse a string when it's expecting a number:

"I wanted a literal integer, but you gave me the value `None`"

Our modem outputs may look similar, but yours may include values the script isn't accounting for. If we could identify the input we should be able to create a check to skip or clean the data.

**Comment by shadowpuck99** [comment]
hmmm.  that's likely the case i suppose.

i'm happy to try and figure out how to do that - but, i also understand if that's out of scope here/you have other things to do with your time!

**Comment by twstokes** [comment]
Sorry it's not a quick fix, but I'll give you some pointers to help troubleshoot:

1. I've added [this new branch](https://github.com/twstokes/arris-scrape/tree/debug) that doesn't use InfluxDB at all and simply outputs the data it would've sent
2. You can pull or download that branch to a clean place on your computer
3. You can skip Docker entirely (since we're just troubleshooting), and first run `pip install -r requirements.txt` to install dependencies
4. Be sure to set up the `config.py` file - you should only have to input your modem URL
5. run `python scrape.py`

Initially you'll most likely see an exception / error output. If things work perfectly, you'll see a bunch of lines with data like:

```
{'measurement': 'upstream', 'tags': {'upstream_id': 'Upstream 3', 'modulation': '64QAM', 'channel_type': 'DOCSIS2.0 (ATDMA)'}, 'fields': {'ucid': 60, 'freq': 36.8, 'power': 42.25, 'symbol_rate': 5120}}
```

Since this is most likely a parsing problem there are two ways this could be fixed:

1. If it's an oversight in how the parser works, it may be fixed at a higher level than a "target"
2. If your device's output has a different structure [than the one I provided](https://github.com/twstokes/arris-scrape/blob/debug/scraper/targets/arris_modem.py), you'll need to create a new target for your particular instance

To summarize, you're probably doing nothing wrong, the script just isn't robust enough to handle many variations of input. 

> ahh, computers are fun....

I agree!

**Comment by shadowpuck99** [comment]
thank you again for spending some time on this - i really do appreciate it....i'll do some experimenting using your tips and see what i can figure out/learn....

**Comment by shadowpuck99** [comment]
i had a few minutes between work projects to give this a quick run....

definitely, based on my limited understanding, that this a parsing issue.....
the first thing i notice is there's no upstream - just need to figure out what's different....
[StatusPageScrape.txt](https://github.com/twstokes/arris-scrape/files/7288101/StatusPageScrape.txt)

**Comment by twstokes** [comment]
Yep, you're on the right track. This is most certainly related to a difference in [this part](https://github.com/twstokes/arris-scrape/blob/master/scraper/targets/arris_modem.py#L122). So, your modem's status page has a slightly different structure.

No pressure to do so, but if you wanted to contribute to open source and possibly help others with your modem in the future, a pull request with a [new target](https://github.com/twstokes/arris-scrape/tree/master/scraper/targets) is welcomed!

**Comment by shadowpuck99** [comment]
ok - so i think i figured this out.

there was a blank <tr> just prior to the first line in the upstream table on my status page.

so, i had to change position()>1 to >2.
[StatusPageScrapeFixed.txt](https://github.com/twstokes/arris-scrape/files/7289299/StatusPageScrapeFixed.txt)

now, to figure out a pull request with the new target script.....

thank you again for the assistance and patience!

**Comment by twstokes** [comment]
Woohoo! That's awesome.

The PR process would be:
1. [Fork this repo](https://docs.github.com/en/get-started/quickstart/fork-a-repo)
2. Make your changes (e.g. [clone this file ](https://github.com/twstokes/arris-scrape/blob/master/scraper/targets/arris_modem.py)and update the model number, software number) and give it a unique name (arris_modem_something.py)
3. [Create the pull request from the fork](https://docs.github.com/en/github/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/creating-a-pull-request-from-a-fork)

Again, no pressure or rush - just if you want to. The script doesn't handle configuring different targets at this time so that'd have to be implemented later.
